src/Microsoft.Azure.WebJobs.ServiceBus/MessagingProvider.cs

```diff
@@ -88,43 +56,23 @@ public virtual MessageProcessor CreateMessageProcessor(string entityPath)
         /// <param name="factory">The <see cref="MessagingFactory"/> to use.</param>
         /// <param name="entityPath">The ServiceBus entity to create a <see cref="MessageReceiver"/> for.</param>
         /// <returns></returns>
-        public virtual MessageReceiver CreateMessageReceiver(string entityPath)
+        public virtual MessageReceiver CreateMessageReceiver(string entityPath, string connectionString)
         {
             if (string.IsNullOrEmpty(entityPath))
             {
                 throw new ArgumentNullException("entityPath");
             }
 
-            // TODO: FACAVAL - Create receiver, can't rely on the factory anymore
-            //MessageReceiver receiver =  factory.CreateMessageReceiver(entityPath);
-            //receiver.PrefetchCount = _config.PrefetchCount;
-            //return receiver;
-
-            return null;
+            return GetOrAddMessageReceiver(entityPath, connectionString);
         }
 
-        /// <summary>
-        /// Gets the connection string for the specified connection string name.
-        /// If no value is specified, the default connection string will be returned.
-        /// </summary>
-        /// <param name="connectionStringName">The connection string name.</param>
-        /// <returns>The ServiceBus connection string.</returns>
-        protected internal string GetConnectionString(string connectionStringName = null)
+        private MessageReceiver GetOrAddMessageReceiver(string entityPath, string connectionString)
         {
-            string connectionString = _config.ConnectionString;
-            if (!string.IsNullOrEmpty(connectionStringName))
-            {
-                connectionString = AmbientConnectionStringProvider.Instance.GetConnectionString(connectionStringName);
-            }
-
-            if (string.IsNullOrEmpty(connectionString))
-            {
-                throw new InvalidOperationException(
-                    string.Format(CultureInfo.InvariantCulture, "Microsoft Azure WebJobs SDK ServiceBus connection string '{0}{1}' is missing or empty.",
-                    "AzureWebJobs", connectionStringName ?? ConnectionStringNames.ServiceBus));
-            }
-
-            return connectionString;
+            return _messageReceiverCache.GetOrAdd(new Tuple<string, string>(entityPath, connectionString),
+                new MessageReceiver(connectionString, entityPath)
```

Comment:
Use the overload that takes a `Func` , which will only get called to create the instance if the key doesn't exist.